### PR TITLE
ensure builddir is not parent of sourcedir

### LINKED ISSUE #11418: Build directory is allowed to be parent directory of project
**Describe the bug**
It is possible, for any project containing meson.build, to configure the parent directory as the build directory. This is easy to do by accident (if you create a build subdirectory and forget to change into it). Critically, 'meson setup --wipe' may then erase the project itself, and any other sibling project.

**To Reproduce**
`meson setup ..`

**Expected behavior**
This should be rejected with an error message. The build directory should not be allowed to be the parent directory of the project containing the `meson.build` file.

Moreover, there seems to be general confusion about the semantics of the directory argument to `meson setup <dir>` since it seems the following are accepted:

`<dir>` is the build directory, current working directory is the project (i.e. contains meson.build)
`<dir>` is the project directory, current working directory is the build directory

While I can understand the occasional convenience of allowing the 2nd form, it doesn't seem to be documented and the fact that it works leads to confusion. The whole bug would be avoided if the 2nd form was rejected, as the documentation implies it should be.

What makes the bug even worse is that `meson setup --wipe ..` will happily erase your project directory, and anything else that happens to be in the same parent directory. It's pretty easy to accidentally configure the parent as build dir i.e.:

1. `meson setup ..` -  whoops, forgot to change into the build subdir first, parent is now configured as build dir. Let's suppose I don't realise, but do see that I wanted a different configuration. Or even that I do realise, but forget to remove the various stuff that meson created in the parent directory so it remains configured as a build directory, then some time later make a similar mistake of forgetting to change into build directory before I do:
2. `meson setup .. -wipe` - whoops, the whole project is gone and so is any other project that happened to be in the parent directory.

I just got hit by this bug. I'm pretty sure I haven't lost anything important, and I realise I'm somewhat to blame, but the fact that this can happen so easily is not good.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
Native build, but probably it doesn't matter.
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Linux.
* what Python version are you using e.g. 3.8.0
~~3.7.5~~ 3.11.1
* what `meson --version`
1.0.0
* what `ninja --version` if it's a Ninja build
N/A, not necessary to run ninja to reproduce.

### PR DESCRIPTION
**Pull request by bruchar1**
Fixes #11418

### CODE REVIEW

File: mesonbuild/msetup.py [hunk 1]
```diff
@@ -160,6 +160,10 @@ def add_vcs_ignore_files(self, build_dir: str) -> None:
 
     def validate_dirs(self, dir1: str, dir2: str, reconfigure: bool, wipe: bool) -> T.Tuple[str, str]:
         (src_dir, build_dir) = self.validate_core_dirs(dir1, dir2)
+
+        if os.path.pardir not in os.path.relpath(src_dir, build_dir):
+            raise SystemExit(f'builddir {build_dir} cannot be a parent of sourcedir {src_dir}')
```
Comment (eli-schwartz) [inline mesonbuild/msetup.py:165]: I would use os.path.commonpath here, and simply check if the common path is the build directory.

Comment (eli-schwartz) [reply mesonbuild/msetup.py:165]: Oh also it seems the Windows tests *really* don't like you using relpath on multiple drives. :)

Comment (bruchar1) [reply mesonbuild/msetup.py]: You are right! Normally, I would use Path.is_relative for that, but it is only available in Python 3.9...

Comment (eli-schwartz) [reply mesonbuild/msetup.py]: Using commonpath should mean you don't need to handle ValueError at all, ntpath.commonpath doesn't raise one here.

That's why I said that it's specifically using relpath that causes drive issues.

Comment (bruchar1) [reply mesonbuild/msetup.py]: Are-you sure?

> Raise ValueError if paths contain both absolute and relative pathnames, the paths are on the different drives or if paths is empty.

Comment (eli-schwartz) [reply mesonbuild/msetup.py]: Hmm, I could swear I tried it last night and it worked. But maybe I'm losing my mind.

Comment (bruchar1) [reply mesonbuild/msetup.py]: `os.path.commonprefix` doesn't throw, but `os.path.commonpath` does. But we really want `commonpath` here, and the cost of exception handling should be neglectable in this context.

Comment (eli-schwartz) [reply mesonbuild/msetup.py]: I agree, but I *thought* I tested commonpath, not commonprefix. Again, I may have simply been losing my mind. :D

**Comment by xclaesse** [comment]
I'm wondering if we should be stricter than that and ensure that builddir is either completely empty or has `meson-private` directory. Currently we check for `meson-private/coredata.dat` to know if that's a builddir but that file only gets created if an initial setup as succeeded. Checking for meson-private directory instead could indicate a partial build.

**Comment by bruchar1** [comment]
We should probably do both. Because if the parent dir contains a  `meson-private` directory, it would still be possible to wipe the source tree....

**Comment by xclaesse** [comment]
I had this commit pending in a WIP branch, made it its own PR now: https://github.com/mesonbuild/meson/pull/11431. It ensures that builddir is either empty or has meson-private subdir.

> We should probably do both. Because if the parent dir contains a `meson-private` directory, it would still be possible to wipe the source tree....

That seems pretty unlikely to happen in practice, but we could create a more explicit stamp file in the meson-private directory to be extra sure it contains a partial build?

**Comment by bruchar1** [comment]
> That seems pretty unlikely to happen in practice, but we could create a more explicit stamp file in the meson-private directory to be extra sure it contains a partial build?

I think it could happen if the builddir was created with an older version of meson allowing  `meson setup ..`. But I would not create a stamp file for that. Checking for `meson-private` is sufficient imho.

**Comment by bruchar1** [comment]
Just rebased to master.

**Comment by xclaesse** [comment]
With https://github.com/mesonbuild/meson/pull/11431 now merged, I think this issue is fixed. Can you please confirm?

**Pull request closed without merging**